File: core/src/main/java/com/pmcl/core/nbt/NbtWriter.java
package com.pmcl.core.nbt;

import java.io.DataOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.file.AtomicMoveNotSupportedException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.Map;
import java.util.zip.GZIPOutputStream;

/**
 * NBT 二进制写入器。
 * <p>
 * 写入 Notchian NBT 格式（大端字节序），默认 gzip 压缩（与 level.dat 兼容）。
 * 落盘采用临时文件 + 原子替换，避免写一半断电/崩溃导致存档损坏。
 */
public final class NbtWriter {

    private NbtWriter() {}

    /**
     * 写入 gzip 压缩的 NBT 文件（level.dat 标准格式）。
     */
    public static void write(NbtTag root, Path file) throws IOException {
        write(root, file, true);
    }

    /**
     * 写入 NBT 文件，可选择是否 gzip。
     * 先写 {@code file.tmp}，再原子替换到目标路径。
     */
    public static void write(NbtTag root, Path file, boolean gzipped) throws IOException {
        Path parent = file.getParent();
        if (parent != null) Files.createDirectories(parent);
        Path tmp = file.resolveSibling(file.getFileName().toString() + ".tmp");
        try {
            try (OutputStream fos = Files.newOutputStream(tmp)) {
                if (gzipped) {
                    try (GZIPOutputStream gz = new GZIPOutputStream(fos);
                         DataOutputStream dos = new DataOutputStream(gz)) {
                        writeRoot(dos, root);
                    }
                } else {
                    try (DataOutputStream dos = new DataOutputStream(fos)) {
                        writeRoot(dos, root);
                    }
                }
            }
            try {
                Files.move(tmp, file,
                        StandardCopyOption.REPLACE_EXISTING,
                        StandardCopyOption.ATOMIC_MOVE);
            } catch (AtomicMoveNotSupportedException e) {
                Files.move(tmp, file, StandardCopyOption.REPLACE_EXISTING);
            }
        } catch (IOException e) {
            try { Files.deleteIfExists(tmp); } catch (IOException ignored) {}
            throw e;
        }
    }

    /**
     * 写入 NBT 到输出流。
     * @param out 输出流
     * @param gzipped 是否 gzip 压缩
     */
    public static void write(NbtTag root, OutputStream out, boolean gzipped) throws IOException {
        OutputStream stream = gzipped ? new GZIPOutputStream(out) : out;
        try (DataOutputStream dos = new DataOutputStream(stream)) {
            writeRoot(dos, root);
        }
    }

    private static void writeRoot(DataOutputStream dos, NbtTag tag) throws IOException {
        dos.writeByte(tag.getType());
        writeString(dos, tag.getName());
        writePayload(dos, tag);
    }

    private static void writePayload(DataOutputStream dos, NbtTag tag) throws IOException {
        switch (tag.getType()) {
            case NbtTag.TYPE_BYTE:
                dos.writeByte(((NbtTag.ByteTag) tag).getValue());
                break;
            case NbtTag.TYPE_SHORT:
                dos.writeShort(((NbtTag.ShortTag) tag).getValue());
                break;
            case NbtTag.TYPE_INT:
                dos.writeInt(((NbtTag.IntTag) tag).getValue());
                break;
            case NbtTag.TYPE_LONG:
                dos.writeLong(((NbtTag.LongTag) tag).getValue());
                break;
            case NbtTag.TYPE_FLOAT:
                dos.writeFloat(((NbtTag.FloatTag) tag).getValue());
                break;
            case NbtTag.TYPE_DOUBLE:
                dos.writeDouble(((NbtTag.DoubleTag) tag).getValue());
                break;
            case NbtTag.TYPE_BYTE_ARRAY: {
                byte[] arr = ((NbtTag.ByteArrayTag) tag).getValue();
                dos.writeInt(arr.length);
                dos.write(arr);
                break;
            }
            case NbtTag.TYPE_STRING:
                writeString(dos, ((NbtTag.StringTag) tag).getValue());
                break;
            case NbtTag.TYPE_LIST: {
                NbtTag.ListTag list = (NbtTag.ListTag) tag;
                dos.writeByte(list.getListType());
                dos.writeInt(list.size());
                for (NbtTag item : list.getItems()) {
                    writePayload(dos, item);
                }
                break;
            }
            case NbtTag.TYPE_COMPOUND: {
                NbtTag.CompoundTag compound = (NbtTag.CompoundTag) tag;
                for (Map.Entry<String, NbtTag> e : compound.getChildren().entrySet()) {
                    dos.writeByte(e.getValue().getType());
                    writeString(dos, e.getKey());
                    writePayload(dos, e.getValue());
                }
                dos.writeByte(NbtTag.TYPE_END);
                break;
            }
            case NbtTag.TYPE_INT_ARRAY: {
                int[] arr = ((NbtTag.IntArrayTag) tag).getValue();
                dos.writeInt(arr.length);
                for (int v : arr) dos.writeInt(v);
                break;
            }
            case NbtTag.TYPE_LONG_ARRAY: {
                long[] arr = ((NbtTag.LongArrayTag) tag).getValue();
                dos.writeInt(arr.length);
                for (long v : arr) dos.writeLong(v);
                break;
            }
            default:
                throw new IOException("未知的 NBT 类型: " + tag.getType());
        }
    }

    private static void writeString(DataOutputStream dos, String s) throws IOException {
        byte[] bytes = s.getBytes(java.nio.charset.StandardCharsets.UTF_8);
        dos.writeShort(bytes.length);
        dos.write(bytes);
    }
}

File: pcmlAndroid/core/src/main/java/com/lash/pmcl/core/nbt/NbtWriter.kt
package com.lash.pmcl.core.nbt

import java.io.DataOutputStream
import java.io.IOException
import java.io.OutputStream
import java.nio.charset.StandardCharsets
import java.nio.file.AtomicMoveNotSupportedException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import java.util.zip.GZIPOutputStream

/**
 * NBT 二进制写入器。
 *
 * 写入 Notchian NBT 格式（大端字节序），默认 gzip 压缩（与 level.dat 兼容）。
 * 落盘采用临时文件 + 原子替换，避免写一半断电/崩溃导致存档损坏。
 *
 * Android 版本：从 Java 移植，使用 sealed class when 替代 instanceof 链。
 */
object NbtWriter {

    /**
     * 写入 gzip 压缩的 NBT 文件（level.dat 标准格式）。
     */
    @Throws(IOException::class)
    fun write(root: NbtTag, file: Path) {
        write(root, file, true)
    }

    /**
     * 写入 NBT 文件，可选择是否 gzip。
     * 先写 `file.tmp`，再原子替换到目标路径。
     */
    @Throws(IOException::class)
    fun write(root: NbtTag, file: Path, gzipped: Boolean) {
        val parent = file.parent
        if (parent != null) Files.createDirectories(parent)
        val tmp = file.resolveSibling(file.fileName.toString() + ".tmp")
        try {
            Files.newOutputStream(tmp).use { fos ->
                if (gzipped) {
                    GZIPOutputStream(fos).use { gz ->
                        DataOutputStream(gz).use { dos ->
                            writeRoot(dos, root)
                        }
                    }
                } else {
                    DataOutputStream(fos).use { dos ->
                        writeRoot(dos, root)
                    }
                }
            }
            try {
                Files.move(
                    tmp, file,
                    StandardCopyOption.REPLACE_EXISTING,
                    StandardCopyOption.ATOMIC_MOVE
                )
            } catch (e: AtomicMoveNotSupportedException) {
                Files.move(tmp, file, StandardCopyOption.REPLACE_EXISTING)
            }
        } catch (e: IOException) {
            try { Files.deleteIfExists(tmp) } catch (ignored: IOException) {}
            throw e
        }
    }

    /**
     * 写入 NBT 到输出流。
     * @param out 输出流
     * @param gzipped 是否 gzip 压缩
     */
    @Throws(IOException::class)
    fun write(root: NbtTag, out: OutputStream, gzipped: Boolean) {
        val stream: OutputStream = if (gzipped) GZIPOutputStream(out) else out
        DataOutputStream(stream).use { dos ->
            writeRoot(dos, root)
        }
    }

    private fun writeRoot(dos: DataOutputStream, tag: NbtTag) {
        dos.writeByte(tag.type)
        writeString(dos, tag.name)
        writePayload(dos, tag)
    }

    private fun writePayload(dos: DataOutputStream, tag: NbtTag) {
        when (tag) {
            is NbtTag.ByteTag -> dos.writeByte(tag.value.toInt())
            is NbtTag.ShortTag -> dos.writeShort(tag.value.toInt())
            is NbtTag.IntTag -> dos.writeInt(tag.value)
            is NbtTag.LongTag -> dos.writeLong(tag.value)
            is NbtTag.FloatTag -> dos.writeFloat(tag.value)
            is NbtTag.DoubleTag -> dos.writeDouble(tag.value)
            is NbtTag.ByteArrayTag -> {
                val arr = tag.value
                dos.writeInt(arr.size)
                dos.write(arr)
            }
            is NbtTag.StringTag -> writeString(dos, tag.value)
            is NbtTag.ListTag -> {
                dos.writeByte(tag.listType)
                dos.writeInt(tag.size)
                for (item in tag.items) {
                    writePayload(dos, item)
                }
            }
            is NbtTag.CompoundTag -> {
                for ((key, value) in tag.children) {
                    dos.writeByte(value.type)
                    writeString(dos, key)
                    writePayload(dos, value)
                }
                dos.writeByte(NbtTag.TYPE_END)
            }
            is NbtTag.IntArrayTag -> {
                val arr = tag.value
                dos.writeInt(arr.size)
                for (v in arr) dos.writeInt(v)
            }
            is NbtTag.LongArrayTag -> {
                val arr = tag.value
                dos.writeInt(arr.size)
                for (v in arr) dos.writeLong(v)
            }
        }
    }

    private fun writeString(dos: DataOutputStream, s: String) {
        val bytes = s.toByteArray(StandardCharsets.UTF_8)
        // NBT 字符串长度字段为 unsigned short；超长时 writeShort 会静默截断但仍写出全部字节，
        // 导致后续标签流损坏。与 NbtReader.MAX_STRING_LEN 对齐，前置拒绝。
        if (bytes.size > 65535) {
            throw IOException("NBT string too long: ${bytes.size} > 65535")
        }
        dos.writeShort(bytes.size)
        dos.write(bytes)
    }
}
